### PR TITLE
Consistent Address prefix with other services (Jörmungandr, Explorer)

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#630 

# Overview

- [x] I have moved "unsafeHumanReadablePartFromText" to the `bech32` package since this is pretty much how we keep on using it :grimacing: We mostly construct 'hrp' from text literals which are correct by construction so the error handling really has little value here.

- [x] Used a single common prefix `addr` regardless of the network discriminant (captured inside the address payload anyway) to be more consistent with Jörmungandr and the explorer.

- [x] I have moved `EncodeAddress` and `DecodeAddress` instances to `Cardano.Wallet.Api.Types` since they really are user-facing encodings that only occurs in the API / CLI.

# Comments

I highly recommend to look at this commit by commit :)

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 1]
```diff
@@ -698,20 +840,44 @@ instance Arbitrary AddressState where
...
-            return (prefix <> addrPayload <> crc)
-        ]
+    arbitrary = (unShowFmt . fst)
+        <$> arbitrary @(ShowFmt Address, Proxy 'Testnet)
+
+instance {-# OVERLAPS #-} Arbitrary (ShowFmt Address, Proxy 'Testnet) where
+    arbitrary = do
+        let proxy = Proxy @'Testnet
+        addr <- ShowFmt <$> frequency
+            [ (10, genAddress 0x83 0x84)
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: no chance to define `0x83` and `0x84` somewhere else and import it as single/group address. What about https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs#L663 And account/mutisig address?

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 2]
```diff
@@ -698,20 +840,44 @@ instance Arbitrary AddressState where
...
+            , (1, genLegacyAddress (30, 100))
+            ]
+        return (addr, proxy)
+
+instance {-# OVERLAPS #-} Arbitrary (ShowFmt Address, Proxy 'Mainnet) where
+    arbitrary = do
+        let proxy = Proxy @'Mainnet
+        addr <- ShowFmt <$> frequency
+            [ (10, genAddress 0x03 0x04)
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: the same comment as above

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 3]
```diff
@@ -698,20 +840,44 @@ instance Arbitrary AddressState where
...
+genAddress single grouped = oneof
+    [ (\bytes -> Address (BS.pack (single:bytes))) <$> vectorOf 32 arbitrary
+    , (\bytes -> Address (BS.pack (grouped:bytes))) <$> vectorOf 64 arbitrary
+    ]
+
+genLegacyAddress :: (Int, Int) -> Gen Address
+genLegacyAddress range = do
+    n <- choose range
+    let prefix = BS.pack
+            [ 130       -- Array(2)
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:868]: maybe here also similar comment as above

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 4]
```diff
@@ -1209,3 +1375,31 @@ atMethod = \case
     DELETE -> delete
     PATCH -> patch
     m -> error $ "atMethod: unsupported method: " <> show m
+
+{-------------------------------------------------------------------------------
+            Generating Golden Test Vectors For Address Encoding
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:1374]: :heart:  for documenting how to generate the stuff

File: specifications/api/swagger.yaml [hunk 5]
```diff
@@ -118,8 +118,11 @@ x-amount: &amount
 
 x-addressId: &addressId
   type: string
-  format: base58
-  example: 2cWKMJemoBam7gg1y5K2aFDhAm5L8fVc96NfxgcGhdLMFTsToNAU9t5HVdBBQKy4iDswL
+  format: base58|bech32
+  examples:
+    - 2cWKMJemoBam7gg1y5K2aFDhAm5L8fVc96NfxgcGhdLMFTsToNAU9t5HVdBBQKy4iDswL
+    - addr1sjck9mdmfyhzvjhydcjllgj9vjvl522w0573ncustrrr2rg7h9azg4cyqd36yyd48t5ut72hgld0fg2xfvz82xgwh7wal6g2xt8n996s3xvu5g
+    - addr1svaervfxt4wc9dycnltq62xwjdcz6l9cl5wnvl9zmltxe20hzwq97350sc8
```
Comment (paweljakubas) [inline specifications/api/swagger.yaml]: :+1:

**Review by paweljakubas — APPROVED** [review]
I know that the PR is in not so small part about code reshuffling, but maybe we can take it as a chance (or during recovery week) to remove magic numbers, defined elsewhere

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 1 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: Right, we could have this is in the `AddressDerivation.Shelley` module itself.

Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: Removed in 566610c8e493fdc4de28078841e74261774be4fe

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 2 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: Removed in 566610c8e493fdc4de28078841e74261774be4fe

**Pull request merged**